src/gtest/test_upgrades.cpp

```diff
@@ -0,0 +1,147 @@
+#include <gtest/gtest.h>
+
+#include "chainparams.h"
+#include "consensus/upgrades.h"
+
+#include <boost/optional.hpp>
+
+TEST(Upgrades, NetworkUpgradeState) {
+    SelectParams(CBaseChainParams::REGTEST);
+    const Consensus::Params& params = Params().GetConsensus();
+
+    // Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT
+    EXPECT_EQ(
+        NetworkUpgradeState(0, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_DISABLED);
+    EXPECT_EQ(
+        NetworkUpgradeState(1000000, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_DISABLED);
+
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, Consensus::NetworkUpgrade::ALWAYS_ACTIVE);
+
+    EXPECT_EQ(
+        NetworkUpgradeState(0, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_ACTIVE);
+    EXPECT_EQ(
+        NetworkUpgradeState(1000000, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_ACTIVE);
+
+    int nActivationHeight = 100;
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, nActivationHeight);
+
+    EXPECT_EQ(
+        NetworkUpgradeState(0, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_PENDING);
+    EXPECT_EQ(
+        NetworkUpgradeState(nActivationHeight - 1, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_PENDING);
+    EXPECT_EQ(
+        NetworkUpgradeState(nActivationHeight, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_ACTIVE);
+    EXPECT_EQ(
+        NetworkUpgradeState(1000000, params, Consensus::UPGRADE_TESTDUMMY),
+        UPGRADE_ACTIVE);
+
+    // Revert to default
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
```

Comment:
This should probably be:
```
BOOST_SCOPE_EXIT() {
    // Revert to default
    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY,
                                   Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
} BOOST_SCOPE_EXIT_END
```

([``BOOST_SCOPE_EXIT``](http://www.boost.org/doc/libs/1_62_0/libs/scope_exit/doc/html/index.html) simulates ``try...finally``, i.e. it runs the given code at scope exit regardless of exceptions. It should be before the first ``UpdateNetworkUpgradeParameters`` call, say at line 12.)

Reply:
This is a Google Test, not a Boost test 😜.

Reply:
I'll do the equivalent.